File: src/engine/router.ts
// ARGUS-P v2 subject-class router — faithful TS port of argus_p/router.py
//
// Classifies a subject from lightweight signals (bio, KB roster presence,
// self-description). Advisory: the analyst can always override. Surfaces every
// class with a positive signal so multi-role subjects are caught.

import { SubjectClass } from "./taxonomy";

const PATTERNS: Record<SubjectClass, RegExp[]> = {
  // Professional capital allocation only. Bare "investing"/"angel" bio talk is
  // retail/KOL noise, not a fund — those words deliberately do NOT score here.
  [SubjectClass.INVESTOR]: [
    /\bventure\b/i, /\bcapital\b/i, /\bVC\b/i, /\bfund\b/i, /\bGP\b/i,
    /\bgeneral partner\b/i, /\blimited partner\b/i, /\bportfolio\b/i,
    /\bangel investor\b/i, /\blaunchpad\b/i, /\baccelerator\b/i,
  ],
  [SubjectClass.FOUNDER]: [
    /\bfounder\b/i, /\bco-?founder\b/i, /\bCEO\b/i, /\bCTO\b/i, /\bbuilding\b/i,
    /\bbuilder\b/i, /\bwe'?re building\b/i, /\bcreator of\b/i, /\bfounded\b/i,
  ],
  // A project/protocol's OWN brand account: an organization, not a person. Uses
  // "we/our", describes one product/token it ships. Distinct from a KOL (who
  // promotes OTHERS' tokens) and a founder (an individual).
  [SubjectClass.PROJECT]: [
    /\bprotocol\b/i, /\bnetwork\b/i, /\bdApp\b/i, /\becosystem\b/i, /\bDAO\b/i,
    /\b(?:prediction|betting|forecasting) market\b/i, /\b(?:decentralized )?exchange\b/i,
    /\bmarketplace\b/i,
    // "Product" is a useful brand-account signal, but not when it is plainly a
    // person's job title. Server routing additionally requires the resolved X
    // profile to link a credible official site before PROJECT can govern.
    /\bproduct\b(?!\s+(?:manager|management|lead|leader|designer|design|engineer|engineering|marketing|marketer|growth|strategy|strategist|ops|operations|at)\b)/i,
    /\bplatform\b/i, /\bthe official\b/i, /\bofficial account\b/i, /\bwe'?re building\b/i,
    /\bour (?:token|protocol|platform|app|mission|community)\b/i, /\b\$[A-Z]{2,6} token\b/i,
    /\bpowered by\b/i, /\bmainnet\b/i, /\btestnet\b/i,
    // Verb-phrase product bios: the account states what its product DOES
    // ("Launch coins on Robinhood", "Trade perps on-chain") without naming a
    // protocol/platform noun. These are brand accounts, not people.
    /\blaunch (?:coins?|tokens?|memecoins?)\b/i,
    /\b(?:create|deploy|mint) (?:a )?(?:coins?|tokens?|memecoins?)\b/i,
    /\btrade (?:perps?|futures|spot|coins?|tokens?|crypto)\b/i,
  ],
  [SubjectClass.ADVISOR]: [
    /\badvisor\b/i, /\badviser\b/i, /\badvisory\b/i, /\bboard member\b/i,
    /\bstrategic advisor\b/i, /\bmentor\b/i,
  ],
  [SubjectClass.AGENCY]: [
    /\bagency\b/i, /\bgrowth\b/i, /\bmarketing\b/i, /\bwe help projects\b/i,
    /\bKOL management\b/i, /\bmarket making\b/i, /\bmarket maker\b/i,
    /\bPR\b/i, /\bservices\b/i, /\bclients\b/i,
  ],
  [SubjectClass.KOL]: [
    /\balpha\b/i, /\bcalls?\b/i, /\btrader\b/i, /\binfluencer\b/i, /\bgems?\b/i,
    /\bdegen\b/i, /\bsignals?\b/i, /\bshill\b/i, /\bcaller\b/i,
  ],
  [SubjectClass.MEMBER]: [
    /\bambassador\b/i, /\bmod\b/i, /\bmoderator\b/i, /\bcommunity\b/i,
    /\bcontributor\b/i,
  ],
};

export interface KBResult {
  kb_hit?: boolean;
  roster_hit?: boolean;
  roster_prices?: string[];
}

export interface RouteResult {
  subject_class: SubjectClass | null;
  applicable_classes: SubjectClass[];
  confidence: string;
  rationale: string;
  scores: Record<SubjectClass, number>;
}

export function classifySubject(
  bio = "",
  kb: KBResult | null = null,
  selfLabel: string | string[] | null = null,
): RouteResult {
  const text = (bio || "").toLowerCase();
  const scores = Object.fromEntries(
    Object.values(SubjectClass).map((c) => [c, 0]),
  ) as Record<SubjectClass, number>;

  for (const cls of Object.keys(PATTERNS) as SubjectClass[]) {
    for (const p of PATTERNS[cls]) {
      if (p.test(text)) scores[cls] += 1;
    }
  }

  const rationale: string[] = [];
  if (kb?.kb_hit) {
    scores[SubjectClass.KOL] += 2;
    rationale.push("present in KOL roster KB (strong KOL prior)");
  }
  if (kb?.roster_hit) {
    scores[SubjectClass.KOL] += 1;
    rationale.push("appears on a paid roster with pricing");
  }

  if (selfLabel != null) {
    const labels = Array.isArray(selfLabel) ? selfLabel : [selfLabel];
    const roles: SubjectClass[] = [];
    for (const lbl of labels) {
      if ((Object.values(SubjectClass) as string[]).includes(lbl)) {
        roles.push(lbl as SubjectClass);
      }
    }
    if (roles.length) {
      return {
        subject_class: roles[0],
        applicable_classes: roles,
        confidence: "operator-set",
        rationale: "roles set explicitly by operator",
        scores,
      };
    }
  }

  const applicable = (Object.keys(scores) as SubjectClass[])
    .filter((c) => scores[c] > 0)
    .sort((a, b) => scores[b] - scores[a]);

  if (!applicable.length) {
    return {
      subject_class: null,
      applicable_classes: [],
      confidence: "none",
      rationale: "no classifying signal; operator must set the role set",
      scores,
    };
  }

  const primary = applicable[0];
  const ordered = Object.values(scores).sort((a, b) => b - a);
  const margin = ordered[0] - (ordered.length > 1 ? ordered[1] : 0);
  const confidence = margin >= 2 ? "high" : "low";
  if (applicable.length > 1) {
    rationale.push("multiple roles detected: " + applicable.join(", "));
  }
  rationale.push(`primary ${primary} (score ${scores[primary]}, margin ${margin})`);
  return {
    subject_class: primary,
    applicable_classes: applicable,
    confidence,
    rationale: rationale.join("; "),
    scores,
  };
}

File: engine_reference/argus_p/router.py
"""
ARGUS-P v2 subject-class router.

Classifies a subject into one of the five classes from lightweight signals
(bio text, KB roster presence, self-description). The router is advisory:
the analyst can always override. Auto-classification is deliberately
conservative and returns its confidence and rationale so the operator can
confirm before the class-specific track runs.
"""

import re
from .taxonomy import SubjectClass

_PATTERNS = {
    SubjectClass.INVESTOR: [
        r'\bventure\b', r'\bcapital\b', r'\bVC\b', r'\bfund\b', r'\bGP\b',
        r'\bgeneral partner\b', r'\blimited partner\b', r'\bportfolio\b',
        r'\bangel\b', r'\binvest(or|ing|ments?)\b', r'\blaunchpad\b', r'\baccelerator\b',
    ],
    SubjectClass.FOUNDER: [
        r'\bfounder\b', r'\bco-?founder\b', r'\bCEO\b', r'\bCTO\b', r'\bbuilding\b',
        r'\bbuilder\b', r'\bwe\'?re building\b', r'\bcreator of\b', r'\bfounded\b',
    ],
    SubjectClass.ADVISOR: [
        r'\badvisor\b', r'\badviser\b', r'\badvisory\b', r'\bboard member\b',
        r'\bstrategic advisor\b', r'\bmentor\b',
    ],
    SubjectClass.AGENCY: [
        r'\bagency\b', r'\bgrowth\b', r'\bmarketing\b', r'\bwe help projects\b',
        r'\bKOL management\b', r'\bmarket making\b', r'\bmarket maker\b',
        r'\bPR\b', r'\bservices\b', r'\bclients\b',
    ],
    SubjectClass.KOL: [
        r'\balpha\b', r'\bcalls?\b', r'\btrader\b', r'\binfluencer\b', r'\bgems?\b',
        r'\bdegen\b', r'\bsignals?\b', r'\bshill\b', r'\bcaller\b',
    ],
    SubjectClass.MEMBER: [
        r'\bambassador\b', r'\bmod\b', r'\bmoderator\b', r'\bcommunity\b',
        r'\bcontributor\b',
    ],
}


def classify_subject(bio: str = "", kb: dict = None, self_label=None) -> dict:
    """Return {'subject_class', 'applicable_classes', 'confidence', 'rationale', 'scores'}.

    Classes are not mutually exclusive. The router surfaces EVERY class with a
    positive signal in `applicable_classes` (ranked), and the strongest as the
    `subject_class` primary, for backward compatibility. A KB roster hit is a
    strong KOL prior. The operator confirms or edits the role set, and can pass
    `self_label` as a single class string or a list of class strings to set the
    roles explicitly.
    """
    text = (bio or "").lower()
    scores = {cls: 0 for cls in SubjectClass}

    for cls, pats in _PATTERNS.items():
        for p in pats:
            if re.search(p, text, re.I):
                scores[cls] += 1

    rationale = []
    if kb and kb.get("kb_hit"):
        scores[SubjectClass.KOL] += 2
        rationale.append("present in KOL roster KB (strong KOL prior)")
    if kb and kb.get("roster_hit"):
        scores[SubjectClass.KOL] += 1
        rationale.append("appears on a paid roster with pricing")

    if self_label is not None:
        labels = self_label if isinstance(self_label, (list, tuple, set)) else [self_label]
        roles = []
        for lbl in labels:
            try:
                roles.append(SubjectClass(lbl))
            except ValueError:
                pass
        if roles:
            return {"subject_class": roles[0], "applicable_classes": roles,
                    "confidence": "operator-set",
                    "rationale": "roles set explicitly by operator", "scores": scores}

    applicable = sorted([c for c, s in scores.items() if s > 0],
                        key=lambda c: scores[c], reverse=True)
    if not applicable:
        return {"subject_class": None, "applicable_classes": [], "confidence": "none",
                "rationale": "no classifying signal; operator must set the role set",
                "scores": scores}

    primary = applicable[0]
    ordered = sorted(scores.values(), reverse=True)
    margin = ordered[0] - (ordered[1] if len(ordered) > 1 else 0)
    confidence = "high" if margin >= 2 else "low"
    if len(applicable) > 1:
        rationale.append("multiple roles detected: " + ", ".join(c.value for c in applicable))
    rationale.append(f"primary {primary.value} (score {scores[primary]}, margin {margin})")
    return {"subject_class": primary, "applicable_classes": applicable,
            "confidence": confidence, "rationale": "; ".join(rationale), "scores": scores}
